pkg/agent/sysadvisor/plugin/qosaware/server/memory_server.go

```diff
@@ -113,6 +114,58 @@ func (ms *memoryServer) populateMetaCache(memoryPluginClient advisorsvc.QRMServi
 	return nil
 }
 
+func (ms *memoryServer) GetAdvice(ctx context.Context, request *advisorsvc.GetAdviceRequest) (*advisorsvc.GetAdviceResponse, error) {
+	_ = ms.emitter.StoreInt64(ms.genMetricsName(metricServerGetAdviceCalled), 1, metrics.MetricTypeNameCount)
+	general.Infof("get advice request: %v", general.ToString(request))
+
+	if err := ms.updateMetaCacheInput(ctx, request); err != nil {
+		general.Errorf("update meta cache failed: %v", err)
+		return nil, fmt.Errorf("update meta cache failed: %w", err)
+	}
+
+	result, err := ms.updateAdvisor()
+	if err != nil {
+		general.Errorf("update advisor failed: %v", err)
+		return nil, fmt.Errorf("update advisor failed: %w", err)
+	}
+	resp := &advisorsvc.GetAdviceResponse{
+		PodEntries:   result.PodEntries,
+		ExtraEntries: result.ExtraEntries,
+	}
+	general.Infof("get advice response: %v", general.ToString(resp))
+	return resp, nil
+}
+
+func (ms *memoryServer) updateMetaCacheInput(ctx context.Context, request *advisorsvc.GetAdviceRequest) error {
+	// lock meta cache to prevent race with cpu server
+	ms.metaCache.Lock()
+	defer ms.metaCache.Unlock()
+
+	var errs []error
+	for podUID, podEntry := range request.Entries {
+		for containerName, containerEntry := range podEntry.Entries {
+			if err := ms.addContainer(containerEntry); err != nil {
+				errs = append(errs, fmt.Errorf("add container %s/%s failed: %w", podUID, containerName, err))
+			}
+		}
+	}
+
+	if err := ms.metaCache.RangeAndDeleteContainer(func(container *types.ContainerInfo) bool {
```

Comment:
whether it will dead lock for it acquiring metaCache lock multiple times?

Reply:
no, line 141 acquires `metaCache.Mutex`, while line 153 acquires `metaCache.podMutex`.